src/playback-watcher.js

```diff
@@ -107,18 +118,71 @@ export default class PlaybackWatcher {
 
     const currentTime = this.tech_.currentTime();
     const buffered = this.tech_.buffered();
+    const atEndBuffer = !buffered.length ||
+      currentTime + Ranges.SAFE_TIME_DELTA >= buffered.end(buffered.length - 1);
+    // If current time is at the end of the final buffered region, then any playback
+    // stall is most likely caused by buffering in a low bandwidth environment. The tech
+    // should fire a `waiting` event in this scenario, but due to browser and tech
+    // inconsistencies. Calling `techWaiting_` here allows us to simulate
+    // responding to a native `waiting` event when the tech fails to emit one.
+
+    if (this.lastRecordedTime === currentTime && atEndBuffer) {
+      if (this.techWaiting_()) {
+        return;
+      }
+
+      const previousBytes = this.stallStartBytes_;
```

Comment:
The biggest part of this change is here. We check if playback is abnormally stalled, if it is we start tracking how many byte transfer changes have happened during this time. If we detect more than 3, ie 3 segments have been downloaded, but not `playing` has occurred. Then we try to blacklist something.